In [53]:
import pandas as pd
import numpy as np
import glob
import os

# Set display options for easier viewing
pd.set_option('display.max_columns', 50)
pd.set_option('display.width', 150)

print("Current working directory:", os.getcwd())
print("\nFiles here:")
for f in os.listdir(os.getcwd()):
    print(f)

Current working directory: /Users/ruchapatwardhan/Desktop/IDX Internship

Files here:
metrics_summary.csv
.DS_Store
listings.ipynb
crmls_sold.py
listings_flagged.csv
CRMLSListing202602.csv
sold-1.ipynb
datascience.ipynb
price_band_evaluation.csv
sold_cleaned.csv
sold_final.csv
listings_final.csv
.gitignore
sold_combined.csv
listings_featured.csv
sold_flagged.csv
CRMLSSold202602.csv
listings_cleaned.csv
listings_combined.csv
.git
sold_featured.csv
crmls_listed.py
raw


In [54]:
with open("crmls_sold.py") as f:
    print(f.read())


import csv
import requests
from datetime import datetime

# Define the API endpoint
url = 'https://api-trestle.corelogic.com/trestle/odata/Property'

# Get token from IDX Exchange secure proxy instead of exposing CoreLogic credentials
auth_endpoint = 'https://idxexchange.com/internal-api/trestle_token.php?key=IDXEXCHANGE2026_CHANGE_THIS'

response = requests.get(auth_endpoint)
response.raise_for_status()

# Parse the response to extract the token

token = response.json().get('access_token')

if token:
    # Define the headers with the token
    headers = {
        'Authorization': f'Bearer {token}'
    }

    # Define the parameters for the API request
    params = {
        '$select': 'BuyerAgentAOR, ListAgentAOR, Flooring,ViewYN,WaterfrontYN,BasementYN,PoolPrivateYN, OriginalListPrice, ListingKey,ListAgentEmail,CloseDate,ClosePrice,ListAgentFirstName,ListAgentLastName,Latitude,Longitude,UnparsedAddress,PropertyType,LivingArea,ListPrice,DaysOnMarket,ListOfficeName,BuyerOfficeName,CoL

In [55]:
df_final = pd.read_csv("sold_final.csv")
print("Shape:", df_final.shape)
print("\nColumns:", df_final.columns.tolist())
df_final.head()

/var/folders/pn/gnz_53jn6wxbzf_7c8wklgf40000gn/T/ipykernel_53422/1897815327.py:1: DtypeWarning: Columns (0,1,4,70,71,72,74,75) have mixed types. Specify dtype option on import or set low_memory=False.
  df_final = pd.read_csv("sold_final.csv")


Shape: (312985, 97)

Columns: ['BuyerAgentAOR', 'ListAgentAOR', 'Flooring', 'ViewYN', 'WaterfrontYN', 'BasementYN', 'PoolPrivateYN', 'OriginalListPrice', 'ListingKey', 'ListAgentEmail', 'CloseDate', 'ClosePrice', 'ListAgentFirstName', 'ListAgentLastName', 'Latitude', 'Longitude', 'UnparsedAddress', 'PropertyType', 'LivingArea', 'ListPrice', 'DaysOnMarket', 'ListOfficeName', 'BuyerOfficeName', 'CoListOfficeName', 'ListAgentFullName', 'CoListAgentFirstName', 'CoListAgentLastName', 'BuyerAgentMlsId', 'BuyerAgentFirstName', 'BuyerAgentLastName', 'AssociationFeeFrequency', 'ListingKeyNumeric', 'MLSAreaMajor', 'CountyOrParish', 'MlsStatus', 'ElementarySchool', 'AttachedGarageYN', 'ParkingTotal', 'BuilderName', 'PropertySubType', 'LotSizeAcres', 'SubdivisionName', 'BuyerOfficeAOR', 'YearBuilt', 'StreetNumberNumeric', 'ListingId', 'BathroomsTotalInteger', 'City', 'BuildingAreaTotal', 'BedroomsTotal', 'ContractStatusChangeDate', 'CoBuyerAgentFirstName', 'PurchaseContractDate', 'ListingContractD

,BuyerAgentAOR,ListAgentAOR,Flooring,ViewYN,WaterfrontYN,BasementYN,PoolPrivateYN,OriginalListPrice,ListingKey,ListAgentEmail,CloseDate,ClosePrice,ListAgentFirstName,ListAgentLastName,Latitude,Longitude,UnparsedAddress,PropertyType,LivingArea,ListPrice,DaysOnMarket,ListOfficeName,BuyerOfficeName,CoListOfficeName,ListAgentFullName,...,BuyerAgencyCompensationType,BuyerAgencyCompensation,latfilled,lonfilled,year_month,rate_30yr_fixed,invalid_close_price,invalid_living_area,invalid_days_on_market,listing_after_close_flag,purchase_after_close_flag,negative_timeline_flag,missing_coords,zero_coords,wrong_longitude,price_ratio,price_per_sqft,year,month,yr_mo,listing_to_contract_days,contract_to_close_days,outlier_close_price,outlier_living_area,outlier_days_on_market
0,SanDiego,SanDiego,NaN,False,NaN,NaN,False,759900.0,522107581,mdarwich12@gmail.com,2024-01-05,815000.0,Michael,Darwich,NaN,NaN,2811 C Avenue,Residential,1974.0,759900.0,33,Berkshire Hathaway HomeServices California Pro...,FOSTER HAMILTON Real Estate,NaN,Michael Darwich,...,NaN,NaN,NaN,NaN,2024-01,6.6425,False,False,False,False,False,False,True,False,False,1.07251,412.867275,2024,1,2024-01,114.0,919.0,False,False,False
1,Mlslistings,Mlslistings,NaN,False,NaN,NaN,NaN,NaN,1079166779,davidmartz@compass.com,2024-01-30,858000.0,David,Martz,NaN,NaN,2199 N Berne Drive,Residential,1995.0,858000.0,0,Compass,NaN,NaN,David Martz,...,NaN,NaN,NaN,NaN,2024-01,6.6425,False,False,False,False,True,False,True,False,False,NaN,430.075188,2024,1,2024-01,188.0,-188.0,False,False,False
2,Southland,Southland,NaN,False,NaN,NaN,False,1890500.0,1075037759,karen.klein@theagencyre.com,2024-01-29,1890500.0,Karen,Klein,34.164376,-118.558749,19524 Winifred Street,Residential,3194.0,1890500.0,0,The Agency,The Agency - Studio City,NaN,Karen Klein,...,NaN,NaN,NaN,NaN,2024-01,6.6425,False,False,False,False,False,False,False,False,False,1.00000,591.891046,2024,1,2024-01,0.0,0.0,False,False,False
3,CoastalMendocino,CoastalMendocino,Tile,True,NaN,NaN,False,1950000.0,1063453216,kira@mendosir.com,2024-01-22,1950000.0,Kira,Meade,39.419080,-123.814842,31451 Bay View Avenue,Residential,2100.0,1950000.0,-36,Mendo Sotheby's International Realty,West Coast Property Management and Sales,NaN,Kira Meade,...,NaN,NaN,NaN,NaN,2024-01,6.6425,False,False,True,False,False,False,False,False,False,1.00000,928.571429,2024,1,2024-01,32.0,34.0,False,False,False
4,Mlslistings,Mlslistings,"Tile,Wood",True,NaN,NaN,NaN,NaN,1061988701,dannyduong@gmail.com,2024-01-31,2340000.0,Danny,Duong,NaN,NaN,3114 Rasmus Circle,Residential,2442.0,2340000.0,0,Infiniti Real Estate,Out of Area Office,NaN,Danny Duong,...,NaN,NaN,NaN,NaN,2024-01,6.6425,False,False,False,False,True,False,True,False,False,NaN,958.230958,2024,1,2024-01,33.0,-33.0,False,False,False


In [56]:
print(df_final['PropertySubType'].value_counts())
print("\nDate range:", df_final['CloseDate'].min(), "to", df_final['CloseDate'].max())

PropertySubType
SingleFamilyResidence    228992
Condominium               54978
Townhouse                 20487
ManufacturedOnLand         3993
Duplex                     1608
StockCooperative           1400
Cabin                       307
Triplex                     211
MixedUse                    111
Quadruplex                   61
BoatSlip                     60
OwnYourOwn                   42
ManufacturedHome             30
MobileHome                   29
Loft                         19
CoOwnership                  14
Timeshare                    11
Studio                        9
Farm                          8
DeededParking                 6
Name: count, dtype: int64

Date range: 2024-01-01 to 2026-01-31


In [57]:
# Filter to the required scope
df = df_final[df_final['PropertySubType'] == 'SingleFamilyResidence'].copy()
print(f"Rows before filter: {len(df_final)}")
print(f"Rows after filter: {len(df)}")

# Convert CloseDate to datetime for time based splitting
df['CloseDate'] = pd.to_datetime(df['CloseDate'])

# Use the most recent month as test, prior months as train
df['close_year_month'] = df['CloseDate'].dt.to_period('M')
most_recent_month = df['close_year_month'].max()
print("\nMost recent month (test set):", most_recent_month)

test_df = df[df['close_year_month'] == most_recent_month].copy()

# Try a training window of X months before the test month, start with X = 6
X = 6
train_months = pd.period_range(end=most_recent_month - 1, periods=X, freq='M')
train_df = df[df['close_year_month'].isin(train_months)].copy()

print(f"\nTrain set (X={X} months): {len(train_df)} rows, {train_df['close_year_month'].min()} to {train_df['close_year_month'].max()}")
print(f"Test set: {len(test_df)} rows")

Rows before filter: 312985
Rows after filter: 228992

Most recent month (test set): 2026-01

Train set (X=6 months): 55024 rows, 2025-07 to 2025-12
Test set: 5867 rows


Week 4

In [58]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score

# Pick a simple set of numeric features for the baseline
features = ['LivingArea', 'BedroomsTotal', 'BathroomsTotalInteger', 'LotSizeSquareFeet', 'YearBuilt']
target = 'ClosePrice'

# Drop rows with missing values in these columns for the baseline
train_model_df = train_df[features + [target]].dropna()
test_model_df = test_df[features + [target]].dropna()

print(f"Train rows after dropping missing: {len(train_model_df)}")
print(f"Test rows after dropping missing: {len(test_model_df)}")

X_train = train_model_df[features]
y_train = train_model_df[target]
X_test = test_model_df[features]
y_test = test_model_df[target]

# Train baseline linear regression
model = LinearRegression()
model.fit(X_train, y_train)

# Evaluate
train_preds = model.predict(X_train)
test_preds = model.predict(X_test)

train_r2 = r2_score(y_train, train_preds)
test_r2 = r2_score(y_test, test_preds)

print(f"\nTrain R²: {train_r2:.4f}")
print(f"Test R²: {test_r2:.4f}")

# Look at coefficients to understand what's driving predictions
coef_df = pd.DataFrame({'Feature': features, 'Coefficient': model.coef_})
print("\nCoefficients:")
print(coef_df)

Train rows after dropping missing: 54055
Test rows after dropping missing: 5737

Train R²: 0.3070
Test R²: 0.3076

Coefficients:
                 Feature    Coefficient
0             LivingArea     346.735945
1          BedroomsTotal  -29138.941694
2  BathroomsTotalInteger  122232.012157
3      LotSizeSquareFeet      -0.000439
4              YearBuilt   -7128.386378


Baseline Linear Regression, Test R² = 0.3076
Features used: LivingArea, BedroomsTotal, BathroomsTotalInteger, LotSizeSquareFeet, YearBuilt
~1,000 train rows and ~130 test rows dropped due to missing values in these features
Negative coefficients on Bedrooms and YearBuilt likely reflect multicollinearity, not causal relationships

Week 5

In [59]:
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor

# Decision Tree
tree_model = DecisionTreeRegressor(max_depth=10, random_state=42)
tree_model.fit(X_train, y_train)

tree_train_preds = tree_model.predict(X_train)
tree_test_preds = tree_model.predict(X_test)

tree_train_r2 = r2_score(y_train, tree_train_preds)
tree_test_r2 = r2_score(y_test, tree_test_preds)

print("Decision Tree:")
print(f"Train R²: {tree_train_r2:.4f}")
print(f"Test R²: {tree_test_r2:.4f}")

# Random Forest
rf_model = RandomForestRegressor(n_estimators=100, max_depth=10, random_state=42, n_jobs=-1)
rf_model.fit(X_train, y_train)

rf_train_preds = rf_model.predict(X_train)
rf_test_preds = rf_model.predict(X_test)

rf_train_r2 = r2_score(y_train, rf_train_preds)
rf_test_r2 = r2_score(y_test, rf_test_preds)

print("\nRandom Forest:")
print(f"Train R²: {rf_train_r2:.4f}")
print(f"Test R²: {rf_test_r2:.4f}")

# Compare all three models side by side
comparison = pd.DataFrame({
    'Model': ['Linear Regression', 'Decision Tree', 'Random Forest'],
    'Train R²': [train_r2, tree_train_r2, rf_train_r2],
    'Test R²': [test_r2, tree_test_r2, rf_test_r2]
})
print("\nModel Comparison:")
print(comparison)

# Feature importance from Random Forest
importance_df = pd.DataFrame({
    'Feature': features,
    'Importance': rf_model.feature_importances_
}).sort_values('Importance', ascending=False)
print("\nRandom Forest Feature Importance:")
print(importance_df)

Decision Tree:
Train R²: 0.4496
Test R²: 0.3510

Random Forest:
Train R²: 0.4796
Test R²: 0.4138

Model Comparison:
               Model  Train R²   Test R²
0  Linear Regression  0.307021  0.307634
1      Decision Tree  0.449632  0.350967
2      Random Forest  0.479615  0.413799

Random Forest Feature Importance:
                 Feature  Importance
0             LivingArea    0.492998
4              YearBuilt    0.279309
3      LotSizeSquareFeet    0.140074
2  BathroomsTotalInteger    0.061050
1          BedroomsTotal    0.026569


In [60]:
# Bed/bath ratio
df['bed_bath_ratio'] = df['BedroomsTotal'] / df['BathroomsTotalInteger'].replace(0, np.nan)

# Property age at time of sale
df['property_age'] = df['CloseDate'].dt.year - df['YearBuilt']

# Quick sanity check on the new features
print("Bed/bath ratio summary:")
print(df['bed_bath_ratio'].describe())

print("\nProperty age summary:")
print(df['property_age'].describe())

# Flag anything odd (e.g. negative age means YearBuilt is after CloseDate, which shouldn't happen)
print("\nRows with negative property age:", (df['property_age'] < 0).sum())
print("Rows with missing bed_bath_ratio:", df['bed_bath_ratio'].isna().sum())

Bed/bath ratio summary:
count    228862.000000
mean          1.514525
std           0.473110
min           0.000000
25%           1.333333
50%           1.500000
75%           1.666667
max           7.500000
Name: bed_bath_ratio, dtype: float64

Property age summary:
count    228920.000000
mean         50.355238
std          26.884705
min          -1.000000
25%          30.000000
50%          51.000000
75%          69.000000
max         249.000000
Name: property_age, dtype: float64

Rows with negative property age: 10
Rows with missing bed_bath_ratio: 130


Week 6

In [61]:
# Flag and inspect the negative age rows
neg_age = df[df['property_age'] < 0][['YearBuilt', 'CloseDate', 'property_age']]
print(neg_age)

        YearBuilt  CloseDate  property_age
136081     2025.0 2024-10-09          -1.0
154286     2025.0 2024-12-30          -1.0
155863     2025.0 2024-12-27          -1.0
213578     2026.0 2025-05-13          -1.0
257346     2026.0 2025-09-09          -1.0
268024     2026.0 2025-09-09          -1.0
270356     2026.0 2025-10-22          -1.0
304513     2026.0 2025-12-23          -1.0
304515     2026.0 2025-12-23          -1.0
304934     2026.0 2025-12-12          -1.0


In [62]:
# Check the oldest properties
print(df[df['property_age'] > 150][['YearBuilt', 'CloseDate', 'property_age', 'City']].head(10))

        YearBuilt  CloseDate  property_age            City
4237       1865.0 2024-01-30         159.0      Wilmington
30851      1860.0 2024-03-05         164.0           Biggs
46914      1870.0 2024-04-12         154.0        Martinez
98462      1859.0 2024-08-22         165.0         Anaheim
104662     1801.0 2024-08-05         223.0        Redlands
132503     1870.0 2024-10-25         154.0         Salinas
133455     1870.0 2024-10-31         154.0       Hollister
156454     1868.0 2024-12-11         156.0        La Verne
160764     1872.0 2025-01-29         153.0   Pacific Grove
162791     1801.0 2025-01-13         224.0  San Bernardino


In [63]:
# Rebuild train/test with the new engineered features included
test_df = df[df['close_year_month'] == most_recent_month].copy()
train_df = df[df['close_year_month'].isin(train_months)].copy()

print(f"Train: {len(train_df)} rows")
print(f"Test: {len(test_df)} rows")

Train: 55024 rows
Test: 5867 rows


In [64]:
!pip install geopandas shapely --quiet


[notice] A new release of pip is available: 24.3.1 -> 26.0.1
[notice] To update, run: /Library/Developer/CommandLineTools/usr/bin/python3 -m pip install --upgrade pip


In [65]:
import geopandas as gpd
from shapely.geometry import Point

school_district_url = "https://data.ca.gov/dataset/california-school-district-areas-2024-25/resource/7dfaf005-58eb-45db-93b1-7aff091b2172/download"
districts = gpd.read_file(school_district_url)
print("District boundaries loaded:", districts.shape)
print(districts.columns.tolist())
districts.head()

District boundaries loaded: (937, 54)
['OBJECTID', 'Year', 'FedID', 'CDCode', 'CDSCode', 'CountyName', 'DistrictNa', 'DistrictTy', 'GradeLow', 'GradeHigh', 'GradeLowCe', 'GradeHighC', 'AssistStat', 'CongressUS', 'SenateCA', 'AssemblyCA', 'UpdateNote', 'EnrollTota', 'EnrollChar', 'EnrollNonC', 'AAcount', 'AApct', 'AIcount', 'AIpct', 'AScount', 'ASpct', 'FIcount', 'FIpct', 'HIcount', 'HIpct', 'PIcount', 'PIpct', 'WHcount', 'WHpct', 'MRcount', 'MRpct', 'NRcount', 'NRpct', 'ELcount', 'ELpct', 'FOScount', 'FOSpct', 'HOMcount', 'HOMpct', 'MIGcount', 'MIGpct', 'SWDcount', 'SWDpct', 'SEDcount', 'SEDpct', 'DistrctAre', 'Shape__Are', 'Shape__Len', 'geometry']


,OBJECTID,Year,FedID,CDCode,CDSCode,CountyName,DistrictNa,DistrictTy,GradeLow,GradeHigh,GradeLowCe,GradeHighC,AssistStat,CongressUS,SenateCA,AssemblyCA,UpdateNote,EnrollTota,EnrollChar,EnrollNonC,AAcount,AApct,AIcount,AIpct,AScount,...,HIpct,PIcount,PIpct,WHcount,WHpct,MRcount,MRpct,NRcount,NRpct,ELcount,ELpct,FOScount,FOSpct,HOMcount,HOMpct,MIGcount,MIGpct,SWDcount,SWDpct,SEDcount,SEDpct,DistrctAre,Shape__Are,Shape__Len,geometry
0,1,2024-25,0601770,0161119,01611190000000,Alameda,Alameda Unified,Unified,PK,12,KG,12,"Differentiated, Year 1",13,09,18,SY 2018-19 Boundary Adjustment/Correction,10771,1479,9292,754,7.0,29,0.3,2673,...,17.7,53,0.5,2880.0,26.7,1684,15.6,320,3.0,910,0.2,18,0.3,77,0.7,0,0.0,1356,12.6,3958,36.7,11.248939,4.755489e+07,56522.982683,"MULTIPOLYGON (((-13606222.82 4540862.699, -136..."
1,2,2024-25,0601860,0161127,01611270000000,Alameda,Albany City Unified,Unified,PK,12,KG,12,General,13,09,15,None,3694,0,3694,142,3.8,9,0.2,1125,...,17.7,2,0.1,899.0,24.3,495,13.4,205,5.5,446,0.1,3,0.0,23,0.6,0,0.0,357,9.7,1184,32.1,1.789984,7.096327e+06,12696.382797,"POLYGON ((-13612893.866 4565099.707, -13612896..."
2,3,2024-25,0604740,0161143,01611430000000,Alameda,Berkeley Unified,Unified,PK,12,KG,12,"Differentiated, Year 1",13,09,15,None,9162,0,9162,1036,11.3,20,0.2,754,...,22.4,20,0.2,3742.0,40.8,1453,15.9,9,0.1,580,0.3,23,0.4,192,2.1,0,0.0,1111,12.1,2686,29.3,10.434329,4.364648e+07,43695.341538,"POLYGON ((-13609482.48 4565074.597, -13609483...."
3,4,2024-25,0607800,0161150,01611500000000,Alameda,Castro Valley Unified,Unified,PK,12,KG,12,"Differentiated, Year 1","13,15","07,09,10","18,20",SY 2018-19 Boundary Adjustment/Correction,9570,0,9570,383,4.0,8,0.1,3253,...,24.5,40,0.4,1783.0,18.6,1003,10.5,327,3.4,1152,0.1,12,0.1,125,1.3,0,0.0,1113,11.6,3728,39.0,66.885571,2.838285e+08,142492.767565,"MULTIPOLYGON (((-13582508.535 4529067.071, -13..."
4,5,2024-25,0612630,0161168,01611680000000,Alameda,Emery Unified,Unified,PK,12,KG,12,General,13,09,15,None,590,0,590,233,39.5,0,0.0,50,...,25.6,5,0.8,64.0,10.8,64,10.8,12,2.0,84,1.2,7,0.2,15,2.5,0,0.0,81,13.7,412,69.8,1.273929,5.363392e+06,13741.272894,"POLYGON ((-13613999.038 4555592.769, -13614126..."


In [66]:
print("District CRS:", districts.crs)

District CRS: EPSG:3857


In [67]:
# Only properties with valid coordinates can be spatially joined
df_geo = df[df['Latitude'].notna() & df['Longitude'].notna()].copy()
print(f"Rows with valid coordinates: {len(df_geo)} out of {len(df)}")

# Convert to a GeoDataFrame using standard lat/long (EPSG:4326)
df_geo['geometry'] = df_geo.apply(lambda row: Point(row['Longitude'], row['Latitude']), axis=1)
gdf_properties = gpd.GeoDataFrame(df_geo, geometry='geometry', crs='EPSG:4326')

# Reproject districts to match, using whatever their actual CRS is
districts_matched = districts.to_crs('EPSG:4326')

# Spatial join: match each property point to the district polygon it falls inside
joined = gpd.sjoin(gdf_properties, districts_matched, how='left', predicate='within')

print("\nJoined shape:", joined.shape)
print("Rows matched to a district:", joined['DistrictNa'].notna().sum())
print("Rows unmatched:", joined['DistrictNa'].isna().sum())

Rows with valid coordinates: 220181 out of 228992

Joined shape: (275679, 155)
Rows matched to a district: 275612
Rows unmatched: 67


check

In [68]:
# Check for duplicate properties in the join (same ListingKey matched to multiple districts)
dup_check = joined['ListingKey'].value_counts()
print("Properties matched to more than one district:", (dup_check > 1).sum())
print("\nExample of a duplicated property:")
example_key = dup_check[dup_check > 1].index[0]
print(joined[joined['ListingKey'] == example_key][['ListingKey', 'DistrictNa', 'DistrictTy']])

Properties matched to more than one district: 55580

Example of a duplicated property:
        ListingKey                 DistrictNa  DistrictTy
83838   1076326017  South Whittier Elementary  Elementary
83838   1076326017        Whittier Union High        High
120544  1076326017  South Whittier Elementary  Elementary
120544  1076326017        Whittier Union High        High
135619  1076326017  South Whittier Elementary  Elementary
135619  1076326017        Whittier Union High        High


In [69]:
# Deduplicate: prefer Unified district type, otherwise take the first match
joined_sorted = joined.sort_values(by='DistrictTy', key=lambda x: x != 'Unified')
joined_dedup = joined_sorted.drop_duplicates(subset='ListingKey', keep='first')

print("Deduplicated shape:", joined_dedup.shape)
print("Original valid coordinate rows:", len(df_geo))

Deduplicated shape: (219980, 155)
Original valid coordinate rows: 220181


Merging

In [70]:
# Merge the district name back into the main dataframe
district_map = joined_dedup.set_index('ListingKey')['DistrictNa']
df['school_district'] = df['ListingKey'].map(district_map)

print("Rows with school_district assigned:", df['school_district'].notna().sum())
print("Rows missing (no coordinates or no match):", df['school_district'].isna().sum())

Rows with school_district assigned: 220117
Rows missing (no coordinates or no match): 8875


Rebuilding

In [71]:
# Rebuild train/test with all Week 6 features included
test_df = df[df['close_year_month'] == most_recent_month].copy()
train_df = df[df['close_year_month'].isin(train_months)].copy()

print(f"Train: {len(train_df)} rows")
print(f"Test: {len(test_df)} rows")

# Updated feature list including Week 6 engineered features
# school_district is categorical, so encode it for now with a simple approach
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()
# Fit on train, handle unseen categories in test by mapping to a placeholder
train_df['school_district_encoded'] = le.fit_transform(train_df['school_district'].astype(str))

# Map test set using the same encoder, unseen districts get -1
district_map_encoded = dict(zip(le.classes_, le.transform(le.classes_)))
test_df['school_district_encoded'] = test_df['school_district'].astype(str).map(district_map_encoded).fillna(-1)

features_v2 = ['LivingArea', 'BedroomsTotal', 'BathroomsTotalInteger', 'LotSizeSquareFeet',
               'YearBuilt', 'bed_bath_ratio', 'property_age', 'school_district_encoded']
target = 'ClosePrice'

train_model_df = train_df[features_v2 + [target]].dropna()
test_model_df = test_df[features_v2 + [target]].dropna()

print(f"\nTrain rows after dropping missing: {len(train_model_df)}")
print(f"Test rows after dropping missing: {len(test_model_df)}")

X_train_v2 = train_model_df[features_v2]
y_train_v2 = train_model_df[target]
X_test_v2 = test_model_df[features_v2]
y_test_v2 = test_model_df[target]

# Retrain Random Forest with the expanded feature set
rf_model_v2 = RandomForestRegressor(n_estimators=100, max_depth=10, random_state=42, n_jobs=-1)
rf_model_v2.fit(X_train_v2, y_train_v2)

rf_v2_train_preds = rf_model_v2.predict(X_train_v2)
rf_v2_test_preds = rf_model_v2.predict(X_test_v2)

rf_v2_train_r2 = r2_score(y_train_v2, rf_v2_train_preds)
rf_v2_test_r2 = r2_score(y_test_v2, rf_v2_test_preds)

print(f"\nRandom Forest v2 (with Week 6 features):")
print(f"Train R²: {rf_v2_train_r2:.4f}")
print(f"Test R²: {rf_v2_test_r2:.4f}")

# Compare old vs new feature set
comparison_v2 = pd.DataFrame({
    'Model': ['Random Forest (Week 5)', 'Random Forest (Week 6, with new features)'],
    'Train R²': [rf_train_r2, rf_v2_train_r2],
    'Test R²': [rf_test_r2, rf_v2_test_r2]
})
print("\nOld vs New Feature Set Comparison:")
print(comparison_v2)

# Updated feature importance
importance_df_v2 = pd.DataFrame({
    'Feature': features_v2,
    'Importance': rf_model_v2.feature_importances_
}).sort_values('Importance', ascending=False)
print("\nUpdated Feature Importance:")
print(importance_df_v2)

Train: 55024 rows
Test: 5867 rows

Train rows after dropping missing: 54039
Test rows after dropping missing: 5735

Random Forest v2 (with Week 6 features):
Train R²: 0.5438
Test R²: 0.4826

Old vs New Feature Set Comparison:
                                       Model  Train R²   Test R²
0                     Random Forest (Week 5)  0.479615  0.413799
1  Random Forest (Week 6, with new features)  0.543773  0.482577

Updated Feature Importance:
                   Feature  Importance
0               LivingArea    0.434305
7  school_district_encoded    0.153032
6             property_age    0.127419
4                YearBuilt    0.115460
3        LotSizeSquareFeet    0.096614
2    BathroomsTotalInteger    0.038821
5           bed_bath_ratio    0.028664
1            BedroomsTotal    0.005683


Week 6 feature engineering (bed/bath ratio, property age, school district spatial join) improved Random Forest Test R² from 0.4138 to 0.4826. School district and property age emerged as strong predictors, validating the geographic feature engineering effort.

Week 7- xgboost installed

In [72]:
from xgboost import XGBRegressor

xgb_model = XGBRegressor(
    n_estimators=200,
    max_depth=6,
    learning_rate=0.1,
    random_state=42,
    n_jobs=-1
)

xgb_model.fit(X_train_v2, y_train_v2)

xgb_train_preds = xgb_model.predict(X_train_v2)
xgb_test_preds = xgb_model.predict(X_test_v2)

xgb_train_r2 = r2_score(y_train_v2, xgb_train_preds)
xgb_test_r2 = r2_score(y_test_v2, xgb_test_preds)

print("XGBoost (initial):")
print(f"Train R²: {xgb_train_r2:.4f}")
print(f"Test R²: {xgb_test_r2:.4f}")

XGBoost (initial):
Train R²: 0.7170
Test R²: 0.6593


In [73]:
from itertools import product

param_grid = {
    'max_depth': [4, 6, 8],
    'learning_rate': [0.05, 0.1],
    'n_estimators': [200, 400]
}

results = []

for depth, lr, n_est in product(param_grid['max_depth'], param_grid['learning_rate'], param_grid['n_estimators']):
    model = XGBRegressor(
        n_estimators=n_est,
        max_depth=depth,
        learning_rate=lr,
        random_state=42,
        n_jobs=-1
    )
    model.fit(X_train_v2, y_train_v2)
    train_preds = model.predict(X_train_v2)
    test_preds = model.predict(X_test_v2)
    train_r2 = r2_score(y_train_v2, train_preds)
    test_r2 = r2_score(y_test_v2, test_preds)
    results.append({
        'max_depth': depth,
        'learning_rate': lr,
        'n_estimators': n_est,
        'train_r2': train_r2,
        'test_r2': test_r2,
        'gap': train_r2 - test_r2
    })

results_df = pd.DataFrame(results).sort_values('test_r2', ascending=False)
print(results_df)

    max_depth  learning_rate  n_estimators  train_r2   test_r2       gap
11          8           0.10           400  0.868931  0.695856  0.173075
9           8           0.05           400  0.809860  0.686281  0.123579
10          8           0.10           200  0.808761  0.684646  0.124115
7           6           0.10           400  0.770876  0.683626  0.087250
5           6           0.05           400  0.716905  0.659545  0.057360
6           6           0.10           200  0.716976  0.659256  0.057720
8           8           0.05           200  0.741319  0.657050  0.084269
3           4           0.10           400  0.673716  0.643844  0.029872
4           6           0.05           200  0.658098  0.620808  0.037291
2           4           0.10           200  0.619024  0.601775  0.017248
1           4           0.05           400  0.618490  0.601033  0.017457
0           4           0.05           200  0.555545  0.546282  0.009263


In [74]:
# Rerun all models fresh to avoid variable name collisions

# Linear Regression
lr_model = LinearRegression()
lr_model.fit(X_train, y_train)
lr_train_r2 = r2_score(y_train, lr_model.predict(X_train))
lr_test_r2 = r2_score(y_test, lr_model.predict(X_test))

# Decision Tree
dt_model = DecisionTreeRegressor(max_depth=10, random_state=42)
dt_model.fit(X_train, y_train)
dt_train_r2 = r2_score(y_train, dt_model.predict(X_train))
dt_test_r2 = r2_score(y_test, dt_model.predict(X_test))

# Random Forest (Week 5 features)
rf5_model = RandomForestRegressor(n_estimators=100, max_depth=10, random_state=42, n_jobs=-1)
rf5_model.fit(X_train, y_train)
rf5_train_r2 = r2_score(y_train, rf5_model.predict(X_train))
rf5_test_r2 = r2_score(y_test, rf5_model.predict(X_test))

# Random Forest (Week 6 features)
rf6_model = RandomForestRegressor(n_estimators=100, max_depth=10, random_state=42, n_jobs=-1)
rf6_model.fit(X_train_v2, y_train_v2)
rf6_train_r2 = r2_score(y_train_v2, rf6_model.predict(X_train_v2))
rf6_test_r2 = r2_score(y_test_v2, rf6_model.predict(X_test_v2))

# XGBoost tuned (Week 7)
xgb_final = XGBRegressor(max_depth=6, learning_rate=0.05, n_estimators=400, random_state=42, n_jobs=-1)
xgb_final.fit(X_train_v2, y_train_v2)
xgb_train_r2 = r2_score(y_train_v2, xgb_final.predict(X_train_v2))
xgb_test_r2 = r2_score(y_test_v2, xgb_final.predict(X_test_v2))

full_comparison = pd.DataFrame({
    'Model': ['Linear Regression', 'Decision Tree', 'Random Forest (Week 5)',
              'Random Forest (Week 6, new features)', 'XGBoost (tuned, Week 7)'],
    'Train R²': [lr_train_r2, dt_train_r2, rf5_train_r2, rf6_train_r2, xgb_train_r2],
    'Test R²': [lr_test_r2, dt_test_r2, rf5_test_r2, rf6_test_r2, xgb_test_r2]
})
print(full_comparison)

                                  Model  Train R²   Test R²
0                     Linear Regression  0.307021  0.307634
1                         Decision Tree  0.449632  0.350967
2                Random Forest (Week 5)  0.479615  0.413799
3  Random Forest (Week 6, new features)  0.543773  0.482577
4               XGBoost (tuned, Week 7)  0.716905  0.659545


XGBoost improved Test R² from 0.4826 (Random Forest, Week 6) to approximately 0.66, the largest single-week performance gain in the project so far. This confirms gradient boosting captures nonlinear relationships in the data that Random Forest and simpler models were missing.

Week 8

In [75]:
# Predictions from the tuned XGBoost model (Week 7)
y_pred = xgb_final.predict(X_test_v2)

# MAPE: Mean Absolute Percentage Error
def mape(y_true, y_pred):
    return np.mean(np.abs((y_true - y_pred) / y_true)) * 100

# MdAPE: Median Absolute Percentage Error (more robust to outliers than MAPE)
def mdape(y_true, y_pred):
    return np.median(np.abs((y_true - y_pred) / y_true)) * 100

test_mape = mape(y_test_v2, y_pred)
test_mdape = mdape(y_test_v2, y_pred)

print(f"Test MAPE: {test_mape:.2f}%")
print(f"Test MdAPE: {test_mdape:.2f}%")
print(f"Test R²: {r2_score(y_test_v2, y_pred):.4f}")

Test MAPE: 16752.94%
Test MdAPE: 15.62%
Test R²: 0.6595


In [76]:
# Build a results dataframe with actual, predicted, and error
results_df = pd.DataFrame({
    'actual': y_test_v2.values,
    'predicted': y_pred
})
results_df['abs_pct_error'] = np.abs((results_df['actual'] - results_df['predicted']) / results_df['actual']) * 100

# Create price bands
bins = [0, 300000, 500000, 750000, 1000000, 1500000, 2000000, np.inf]
labels = ['<300k', '300k-500k', '500k-750k', '750k-1M', '1M-1.5M', '1.5M-2M', '2M+']
results_df['price_band'] = pd.cut(results_df['actual'], bins=bins, labels=labels)

# Summarize by price band
band_summary = results_df.groupby('price_band', observed=True).agg(
    count=('actual', 'size'),
    mean_mape=('abs_pct_error', 'mean'),
    median_mape=('abs_pct_error', 'median')
).round(2)
print(band_summary)

            count  mean_mape  median_mape
price_band                               
<300k         156  615105.68       102.59
300k-500k     844      40.02        26.62
500k-750k    1587      21.71        15.18
750k-1M      1326      16.72        12.88
1M-1.5M      1162      14.75        11.94
1.5M-2M       480      18.34        16.46
2M+           180      29.35        29.69


In [77]:
# Investigate the worst offenders in the <300k band
low_band = results_df[results_df['price_band'] == '<300k'].sort_values('abs_pct_error', ascending=False)
print(low_band.head(10))

# Check for suspiciously low actual prices
print("\nRows with actual < $10,000:")
print(results_df[results_df['actual'] < 10000])

         actual     predicted  abs_pct_error price_band
4063       1.75  1.678800e+06   9.593134e+07      <300k
1283   57000.00  9.080304e+05   1.493036e+03      <300k
3749   75500.00  1.101782e+06   1.359314e+03      <300k
4530   31000.00  4.096779e+05   1.221542e+03      <300k
4723   83000.00  9.592789e+05   1.055758e+03      <300k
3308   55000.00  5.277306e+05   8.595102e+02      <300k
1267  165000.00  1.332898e+06   7.078170e+02      <300k
410   100000.00  7.313472e+05   6.313472e+02      <300k
3345  140000.00  9.021422e+05   5.443873e+02      <300k
3106  115000.00  7.331326e+05   5.375066e+02      <300k

Rows with actual < $10,000:
      actual   predicted  abs_pct_error price_band
4063    1.75  1678800.25   9.593134e+07      <300k


fixing

In [78]:
# Recalculate MAPE excluding clear data errors (e.g. price below a reasonable floor)
clean_results = results_df[results_df['actual'] >= 50000].copy()

clean_mape = clean_results['abs_pct_error'].mean()
clean_mdape = clean_results['abs_pct_error'].median()

print(f"Rows excluded as likely data errors: {len(results_df) - len(clean_results)}")
print(f"Cleaned MAPE (actual >= $50,000): {clean_mape:.2f}%")
print(f"Cleaned MdAPE (actual >= $50,000): {clean_mdape:.2f}%")

Rows excluded as likely data errors: 2
Cleaned MAPE (actual >= $50,000): 25.39%
Cleaned MdAPE (actual >= $50,000): 15.61%


Summary

In [79]:
metrics_summary = pd.DataFrame({
    'Metric': ['R²', 'MAPE (raw)', 'MAPE (cleaned, excl. data errors)', 'MdAPE'],
    'Value': [f"{r2_score(y_test_v2, y_pred):.4f}", f"{test_mape:.2f}%", f"{clean_mape:.2f}%", f"{test_mdape:.2f}%"]
})
print(metrics_summary)

# Save the price band breakdown and metrics summary for your documentation
band_summary.to_csv('price_band_evaluation.csv')
metrics_summary.to_csv('metrics_summary.csv', index=False)
print("\nSaved metrics_summary.csv and price_band_evaluation.csv")

                              Metric      Value
0                                 R²     0.6595
1                         MAPE (raw)  16752.94%
2  MAPE (cleaned, excl. data errors)     25.39%
3                              MdAPE     15.62%

Saved metrics_summary.csv and price_band_evaluation.csv


- Test R² = 0.6595, the tuned XGBoost model explains about 66% of the variance in close price
- MdAPE = 15.61%, the most reliable headline metric, predictions are typically within about 16% of actual price
- Raw MAPE was severely distorted by 2 data entry errors (one row had an actual close price of $1.75); after excluding those, cleaned MAPE = 25.39%
- The gap between MAPE (25.39%) and MdAPE (15.61%) is itself worth noting, it shows a subset of predictions have large errors that pull the average up, worth investigating further, likely concentrated in the very low and very high price bands as the earlier breakdown showed
- Performance by price band: strongest in the 750k-1.5M range (MdAPE ~12%), weakest at the extremes, under 300k (after removing bad data) and 2M+ (MdAPE ~30%)